pkg/sql/sem/tree/table_name.go

```diff
@@ -94,33 +94,73 @@ type TableNameReference interface {
 
 // TableName corresponds to the name of a table in a FROM clause,
 // INSERT or UPDATE statement (and possibly other places).
+//
+// Note: user code out of this package should probably not construct
+// instances of this directly, and instead use the NewTableName /
+// MakeTableName functions underneath.
 type TableName struct {
-	PrefixName   Name
-	DatabaseName Name
-	TableName    Name
-
-	// OmitDBNameDuringFormatting, when set to true, causes the
-	// String()/Format() methods to omit the database name even if one
-	// is set. This is used to ensure that pretty-printing
-	// a TableName normalized from a parser input yields back
-	// the original syntax.
-	OmitDBNameDuringFormatting bool
-
-	// PrefixOriginallySpecified indicates whether a prefix was
-	// explicitly indicated in the input syntax.
-	PrefixOriginallySpecified bool
+	CatalogName Name
+	SchemaName  Name
+
+	TableName Name
+
+	// ExplicitCatalog is true iff the catalog was explicitly specified
+	// or it needs to be rendered during pretty-printing.
+	ExplicitCatalog bool
+	// ExplicitSchema is true iff the schema was explicitly specified
+	// or it needs to be rendered during pretty-printing.
+	ExplicitSchema bool
+}
+
+// MakeTableName creates a new table name qualified with just a schema.
+func MakeTableName(db, tbl Name) TableName {
+	return TableName{
+		SchemaName:     db,
+		TableName:      tbl,
+		ExplicitSchema: true,
+	}
+}
+
+// NewTableName creates a new qualified table name.
+func NewTableName(db, tbl Name) *TableName {
+	tn := MakeTableName(db, tbl)
+	return &tn
+}
+
```

Comment:
This form is weird. Can we reconsider it now that we're switching the semantics of our naming hierarchy? Remind me why we need it again?